### PR TITLE
[dist_optim] fix the bug of none grads on functional optimizers

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62249

parameter and grads passed to torch.optim.functional should always match, we should skip the parameters that have none gradients to avoid the size mismatch

Differential Revision: [D29929653](https://our.internmc.facebook.com/intern/diff/D29929653/)

### CODE REVIEW

File: torch/testing/_internal/distributed/rpc/dist_optimizer_test.py [hunk 1]
```diff
@@ -209,3 +209,67 @@ def test_dist_optim(self):
...
+        loss.backward()
+        local_optim.step()
+
+        # distributed version
+        owner1 = "worker%d" % ((self.rank + 1) % self.world_size)
+        owner2 = "worker%d" % ((self.rank + 2) % self.world_size)
+
+        remote_module1 = rpc.remote(owner1, MyModule)
+        remote_module2 = rpc.remote(owner2, MyModule, args=(False,))
+        remote_param1 = remote_method(MyModule.get_w, remote_module1)
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:241]: nit: this can be simplified to `remote_module1.remote().get_w()`?

File: torch/testing/_internal/distributed/rpc/dist_optimizer_test.py [hunk 2]
```diff
@@ -209,3 +209,67 @@ def test_dist_optim(self):
...
+        remote_param2 = remote_method(MyModule.get_w, remote_module2)
+
+        old_w1_remote = remote_param1.to_here()
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:244]: `old_w1_remote` not used?

File: torch/testing/_internal/distributed/rpc/dist_optimizer_test.py [hunk 3]
```diff
@@ -209,3 +209,67 @@ def test_dist_optim(self):
...
+
+        dist_optim = DistributedOptimizer(
+            optim.SGD, [remote_param1, remote_param2], lr=0.05
+        )
+
+        with dist_autograd.context() as context_id:
+            g_cpu.manual_seed(0)
+            t1 = torch.rand((3, 3), requires_grad=True, generator=g_cpu)
+            t2 = torch.rand((3, 3), requires_grad=True, generator=g_cpu)
+            output1 = rpc_async_method(MyModule.forward, remote_module1, t2)
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:258]: nit: this can be `remote_module1.rpc_async().forward(t2)`?

File: torch/testing/_internal/distributed/rpc/dist_optimizer_test.py [hunk 4]
```diff
@@ -209,3 +209,67 @@ def test_dist_optim(self):
...
+            output2 = rpc_async_method(MyModule.forward, remote_module2, output1.wait())
+            loss = torch.add(output2.wait(), t1)
+
+            dist_autograd.backward(context_id, [loss.sum()])
+            dist_optim.step(context_id)
+
+            new_w1 = rpc_async_method(MyModule.get_w, remote_module1).wait()
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:265]: ditto

**Review by mrshenli — APPROVED** [review]
LGTM! Left some minor comments.

**Review by mrshenli — COMMENTED** [review]
BTW, looks like we need the same fix for `functional_rprop.py` as well

https://github.com/pytorch/pytorch/blob/a3be2ecc3aee3b0ea0a27f803b8e6ac3245c6789/torch/distributed/optim/functional_rprop.py#L40-L71

**Comment by andwgu** [comment]
Thanks for patching this quickly!
